agent/installer/os_detector.go

```diff
@@ -15,44 +15,69 @@ const (
 	osNotDetected = "could not detect OS correctly"
 )
 
-// oSDetector contains all the logic for detecting the OS version.
-type osDetector struct {
-	cachedNormalizedOS string
+// OSDetector contains all the logic for detecting the OS version.
+type OSDetector struct {
+	os   string
+	ver  string
+	arch string
 }
 
 // Detect returns the os info in normalized format.
 // The format is as follows: <os>_<ver>_<arch>
 // Example with Ubuntu 21.04.3 64bit: Ubuntu_20.04.3_x64
-func (osd *osDetector) Detect() (string, error) {
-	return osd.DetectByHostnamectl(func() (string, error) { return osd.getHostnamectl() })
+func (osd *OSDetector) Detect() (string, error) {
+	return osd.GetNormalizedOS(func() (string, error) { return osd.getHostnamectl() })
 }
 
-// DetectByHostnamectl is a helper method to enable testing of detect with mock methods.
-func (osd *osDetector) DetectByHostnamectl(f func() (string, error)) (string, error) {
-	if osd.cachedNormalizedOS != "" {
-		return osd.cachedNormalizedOS, nil
+// GetOS returns the os name along with the version
+// Example: Ubuntu 21.04.3
+func (osd *OSDetector) GetOS() (string, error) {
+	return osd.GetOSNameWithVersion(func() (string, error) { return osd.getHostnamectl() })
+}
+
+// GetNormalizedOS detects and converts the detected os into format <os>_<ver>_<arch>
+func (osd *OSDetector) GetNormalizedOS(f func() (string, error)) (string, error) {
+	if osd.os != "" && osd.ver != "" && osd.arch != "" {
+		return osd.normalizeOSName(), nil
+	}
+
+	if err := osd.DetectByHostnamectl(f); err != nil {
+		return "", err
+	}
+
+	return osd.normalizeOSName(), nil
+}
+
+// GetOSNameWithVersion detects and converts the detected os into format <os> <ver>
+func (osd *OSDetector) GetOSNameWithVersion(f func() (string, error)) (string, error) {
+	if osd.os != "" && osd.ver != "" {
```

Comment:
Just a small clarification: These `os` and `ver` values can also be fetched during reconciliation and that's why we do this optimization so as not to run `hostnamectl` again, right?

Reply:
Yes if the struct OSDetector is already contains those fields it skips calling `hostnamectl`. But the function declares the object each time it's called so the intention is not fulfilled. I guess we should declare the object at global level in the reconciler, right?

Reply:
> I guess we should declare the object at global level in the reconcile

Yeah, makes sense!